sympy/functions/combinatorial/numbers.py

```diff
@@ -733,13 +734,20 @@ class euler(Function):
 
     @classmethod
     def eval(cls, m):
-        if m.is_odd:
-            return S.Zero
-        if m.is_Integer and m.is_nonnegative:
-            from mpmath import mp
-            m = m._to_mpmath(mp.prec)
-            res = mp.eulernum(m, exact=True)
-            return Integer(res)
+        if m.is_integer:
+            if fuzzy_or([m.is_positive, m.is_negative]) is None:
+                return
+            elif m.is_odd and m >= 1:
+                return S.Zero
+            elif m.is_nonnegative:
+                from mpmath import mp
+                m = m._to_mpmath(mp.prec)
+                res = mp.eulernum(m, exact=True)
+                return Integer(res)
+        if m.is_negative:
+            raise ValueError('Euler numbers are not defined on negative numbers')
+        elif m.is_rational or m.is_positive is False:
```

Comment:
@jksuom Now, that's looks good. ```is_rational``` is valid because i cover ```positive```, ```negative``` and zero in above condition and we need to return ```ValueError``` except those numbers.